### PR TITLE
Pi 3 support [WIP]

### PR DESCRIPTION
**Pull request by StaticSaga**
# Roadblocks
- [ ] actually test on real hardware
- [x] fix bug on ReleaseSmall 

# Features to add in a later PR:
- reading kernel file from SD card
- don't hardcode the resolution

### CODE REVIEW

File: build.zig [hunk 1]
```diff
@@ -178,6 +196,11 @@ pub fn build_blob(b: *Builder, arch: builtin.Arch, target_name: []const u8, path
   return blob(b, elf, .Padded);
 }
 
+pub fn build_pi3_kernel8(b: *Builder, arch: builtin.Arch, target_name: []const u8, path_prefix: []const u8) !*TransformFileCommandStep {
+  const elf = try build_elf(b, arch, target_name, path_prefix);
+  return pi3_kernel8(b, elf);
```
Comment (N00byEdge) [inline build.zig]: Just call `blob(b, elf, .NotPadded)` and scrap `pi3_kernel8`? Or am I missing something?

File: src/lib/pmm.zig [hunk 2]
```diff
@@ -76,10 +76,10 @@ pub fn switch_state(new_state: pmm_state) void {
   var current_base = sabaton.near("pmm_head").read(u64);
   current_base += page_size - 1;
   current_base &= ~(page_size - 1);
-
+  
   const eff_idx = @as(usize, @enumToInt(current_state)) * 3;
   const current_entry = sabaton.near("dram_base").addr(u64);
-
+  current_entry[eff_idx + 0] = dram;
```
Comment (N00byEdge) [inline src/lib/pmm.zig]: Are you sure about this? Doesn't this set every memmap entry to start at the dram base?

**Comment by circutrider21** [comment]
Reading from the sdcard is not required since one could specify in config.txt a file to be loaded into memory at a fixed address like so
`initramfs <file_name> 0x00800000`
or if you want the image placed in memory after the bootloader, use the followkernel specifier like this
`initramfs <file_name> followkernel`

Therefore, you can load the kernel without needing to read from the sdcard
[reference link](https://www.raspberrypi.org/documentation/computers/config_txt.html#boot-options)

**Review by N00byEdge — APPROVED** [review]

**Comment by N00byEdge** [comment]
> As with ramfsfile, newer firmwares allow the loading of multiple files by comma-separating their names.

oooh, this sounds exactly what we'd wanna do to avoid putting fs code in sabaton. At some point you should wonder if it's worth it though...

**Comment by StaticSaga** [comment]
The code already assumes the kernel is loaded at address 0x200000 (loader device on QEMU, and config.txt on real hardware), but I think this is suboptimal, as it requires fiddling with config.txt. A polling Broadcom sdhci driver is not that complex, but I would like to verify that the current code works on real hardware first.

**Comment by N00byEdge** [comment]
I would agree about that to, when adjusting the pmm base for sabaton, I had to make the assumption that the kernel doesn't exceed the size of 16M, feels like an unnecessary assumption to make, would love to get the file size. Also wastes the full 16M permanently, regardless of kernel size.

**Comment by StaticSaga** [comment]
It's not a big issue, in my opinion. Not even GRUB exceeds 16M!

**Comment by N00byEdge** [comment]
Yeah, but I could see someone embedding some GNU/Bloat initrd in their kernel exceeding that.

**Pull request merged**